### PR TITLE
Rework climate traits

### PR DESCRIPTION
**Pull request by OttoWinter**
# What does this implement/fix? 

The whole `ClimateTraits` class looked very wonky with the hundreds of boolean flags. Instead, refactor it to `std::set` instances. Old methods are kept just in case a custom integration uses it, but marked as deprecated.

Also fixes the `ClimatePreset` enum 0 value mentioned in https://github.com/esphome/aioesphomeapi/pull/24#pullrequestreview-687339203 (new 0 value is `CLIMATE_PRESET_NONE`, indicating that the field is not set).

`away` is also deprecated, as it _should_ now be done with `PRESET_AWAY`.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>
closes #1944

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by Expaso** [comment]
I am so exited these are being pulled off right now!  Just when I ripped out the WIFI module of my Panasonic AC, and replaced it with a nice ESPHome device! 
What is ESPHome a wonderfull platform! 😃

File: esphome/components/midea_ac/midea_climate.cpp [hunk 1]
```diff
@@ -167,24 +167,38 @@ climate::ClimateTraits MideaAC::traits() {
...
-  traits.set_supports_dry_mode(true);
-  traits.set_supports_heat_mode(true);
-  traits.set_supports_fan_only_mode(true);
-  traits.set_supports_fan_mode_auto(true);
-  traits.set_supports_fan_mode_low(true);
-  traits.set_supports_fan_mode_medium(true);
-  traits.set_supports_fan_mode_high(true);
+  traits.set_supported_modes({
+      climate::CLIMATE_MODE_OFF,
+      climate::CLIMATE_MODE_AUTO,
```
Comment (jesserockz) [inline esphome/components/midea_ac/midea_climate.cpp:172]:
```suggestion
      climate::CLIMATE_MODE_HEAT_COOL,
```

File: esphome/components/thermostat/thermostat_climate.cpp [hunk 2]
```diff
@@ -78,27 +79,49 @@ void ThermostatClimate::control(const climate::ClimateCall &call) {
...
-  traits.set_supports_fan_mode_high(this->supports_fan_mode_high_);
-  traits.set_supports_fan_mode_middle(this->supports_fan_mode_middle_);
-  traits.set_supports_fan_mode_focus(this->supports_fan_mode_focus_);
-  traits.set_supports_fan_mode_diffuse(this->supports_fan_mode_diffuse_);
-  traits.set_supports_swing_mode_both(this->supports_swing_mode_both_);
-  traits.set_supports_swing_mode_horizontal(this->supports_swing_mode_horizontal_);
-  traits.set_supports_swing_mode_off(this->supports_swing_mode_off_);
-  traits.set_supports_swing_mode_vertical(this->supports_swing_mode_vertical_);
+  if (supports_auto_)
+    traits.add_supported_mode(climate::CLIMATE_MODE_HEAT_COOL);
```
Comment (jesserockz) [inline esphome/components/thermostat/thermostat_climate.cpp:83]:
```suggestion
  if (supports_heat_cool_)
    traits.add_supported_mode(climate::CLIMATE_MODE_HEAT_COOL);
```

File: esphome/components/midea_ac/midea_climate.cpp (hunk 1 above)
Comment (OttoWinter) [reply esphome/components/midea_ac/midea_climate.cpp:172]: Thanks!

File: esphome/components/thermostat/thermostat_climate.cpp (hunk 2 above)
Comment (OttoWinter) [reply esphome/components/thermostat/thermostat_climate.cpp]: Nice catch! Thanks!

**Review by jesserockz — APPROVED** [review]

**Comment by OttoWinter** [comment]
Thanks for the review! I looked over the code again and found some other small issues :face_with_head_bandage: but should be fixed now :+1: 

(also undid the change with adding `PRESET_NONE` to be more consistent with other API enums)

**Review by jesserockz — APPROVED** [review]
Cool yeah. I think HOME does make a bit more sense as the default anyway

**Pull request merged**